plugins/inputs/openstack/openstack.go

```diff
@@ -0,0 +1,1059 @@
+	if err != nil {
+		return fmt.Errorf("unable to create Newclient for OpenStack endpoint %v", err)
+	}
+
+	tlsCfg, err := o.ClientConfig.TLSConfig()
+	if err != nil {
+		return err
+	}
+
+	provider.HTTPClient = http.Client{
+		Transport: &http.Transport{
+			TLSClientConfig: tlsCfg,
+			Proxy:           http.ProxyFromEnvironment,
+		},
+		Timeout: time.Duration(o.Timeout),
+	}
+
+	if err := openstack.Authenticate(provider, authOption); err != nil {
+		return fmt.Errorf("unable to authenticate OpenStack user %v", err)
+	}
+
+	// Create required clients and attach to the OpenStack struct
+	if o.identity, err = openstack.NewIdentityV3(provider, gophercloud.EndpointOpts{}); err != nil {
+		return fmt.Errorf("unable to create V3 identity client %v", err)
+	}
+
+	if err := o.gatherServices(); err != nil {
+		o.Log.Warnf("failed to get resource openstack services %v", err)
+	}
+
+	if o.compute, err = openstack.NewComputeV2(provider, gophercloud.EndpointOpts{}); err != nil {
+		return fmt.Errorf("unable to create V2 compute client %v", err)
+	}
+
+	// Create required clients and attach to the OpenStack struct
+	if o.network, err = openstack.NewNetworkV2(provider, gophercloud.EndpointOpts{}); err != nil {
+		return fmt.Errorf("unable to create V2 network client %v", err)
+	}
+
+	// The Orchestration service is optional
+	if o.ContainsService(Orchestration) {
+		if o.stack, err = openstack.NewOrchestrationV1(provider, gophercloud.EndpointOpts{}); err != nil {
+			return fmt.Errorf("unable to create V1 stack client %v", err)
+		}
+	}
+
+	// The Cinder volume storage service is optional
+	if o.ContainsService(volumeV2Service) {
+		if o.volume, err = openstack.NewBlockStorageV2(provider, gophercloud.EndpointOpts{}); err != nil {
+			return fmt.Errorf("unable to create V2 volume client %v", err)
+		}
+	}
+
+	o.gather_servers = false
+
+	return nil
+}
+
+// Gather gathers resources from the OpenStack API and accumulates metrics.  This
+// implements the Input interface.
+func (o *OpenStack) Gather(acc telegraf.Accumulator) error {
+
+	// Gather resources.  Note service harvesting must come first as the other
+	// gatherers are dependant on this information.
+	gatherers := map[string]func() error{
+		"projects":      o.gatherProjects,
+		"hypervisors":   o.gatherHypervisors,
+		"flavors":       o.gatherFlavors,
+		"servers":       o.gatherServers,
+		"volumes":       o.gatherVolumes,
+		"storage_pools": o.gatherStoragePools,
+		"subnets":       o.gatherSubnets,
+		"ports":         o.gatherPorts,
+		"networks":      o.gatherNetworks,
+		"aggregates":    o.gatherAggregates,
+		"nova_services": o.gatherNovaServices,
+		"agents":        o.gatherAgents,
+		"stacks":        o.gatherStacks,
+	}
+
+	for _, service := range o.EnabledServices {
+		if service != "services" {
+			gatherer := gatherers[service]
+			if err := gatherer(); err != nil {
+				o.Log.Warnf("failed to get resource %q %v", service, err)
+			}
+		}
```

Comment:
Please invert the condition to save an indention level. Furthermore, I guess it's an error if you cannot get a service so use `acc.AddError()`:
```suggestion
		if service == "services" {
			continue
		}
		
		gatherer := gatherers[service]
		if err := gatherer(); err != nil {
			acc.AddError(fmt.Errorf("failed to get resource %q %v", service, err))
		}
```

Reply:
I am assuming that acc.AddError will not terminate the process/flow and collect metrics for rest of the configured services. Is my assumption right?

Reply:
Yes your assumption is correct. That's the intended use-case for `acc.AddError()`.

Reply:
Thanks for confirming.